proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+  public var count: Int { get }
+
+  /// A Boolean value indicating whether the span is empty.
+  public var isEmpty: Bool { get }
+
+  /// The indices that are valid for subscripting the span, in ascending
+  /// order.
+  public var indices: Range<Int> { get }
+
+  /// Traps if `offset` is not a valid offset into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - position: an offset to validate
+  public func boundsCheckPrecondition(_ offset: Int)
+
+  /// Traps if `bounds` is not a valid range of offsets into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - offsets: a range of offsets to validate
+  public func boundsCheckPrecondition(_ offsets: Range<Int>)
+}
+```
+
+##### Index validation utiliities:
+
+Every time `RawSpan` uses an index or an integer offset, it checks for their validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:
+
+```swift
+extension RawSpan {
+  /// Traps if `position` is not a valid index for this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - position: an Index to validate
+  public boundsCheckPrecondition(_ position: Index)
+
+  /// Traps if `bounds` is not a valid range of indices for this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - bounds: a range of indices to validate
+  public boundsCheckPrecondition(_ bounds: Range<Index>)
+
+  /// Traps if `offset` is not a valid offset into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - offset: an offset to validate
+  public boundsCheckPrecondition(offset: Int)
+
+  /// Traps if `offsets` is not a valid range of offsets into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - offsets: a range of offsets to validate
+  public boundsCheckPrecondition(offsets: Range<Int>)
+}
+```
+
+##### Accessing subranges of elements:
+
+Similarly to `Span`, `RawSpan` does not support slicing in the style of `Collection`. It supports a similar set of `extracting()` functions as `Span`:
+
+```swift
+extension RawSpan {
+  /// Constructs a new span over the bytes within the supplied range of
+  /// positions within this span.
+  ///
+  /// The returned span's first byte is always at offset 0; unlike buffer
+  /// slices, extracted spans do not generally share their indices with the
+  /// span from which they are extracted.
+  ///
+  /// - Parameter bounds: A valid range of positions. Every position in
+  ///     this range must be within the bounds of this `RawSpan`.
+  ///
+  /// - Returns: A `Span` over the bytes within `bounds`
+  public func extracting(_ bounds: Range<Int>) -> Self
+
+  /// Constructs a new span over the bytes within the supplied range of
+  /// positions within this span.
+  ///
+  /// The returned span's first byte is always at offset 0; unlike buffer
+  /// slices, extracted spans do not generally share their indices with the
+  /// span from which they are extracted.
+  ///
+  /// This function does not validate `bounds`; this is an unsafe operation.
```

Comment:
General note: We need to avoid guaranteeing that no validation will take place, as we will still need these to fully validate their inputs on certain platforms and/or configurations. Such promises should be weakened accordingly.

```suggestion
  /// This function may not always validate `bounds`; this is an unsafe operation.
```